rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,363 @@
+    - Many packages fit multiple categories, leading to multiple locations to search through instead of one
+  - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)
+- This breaks `builtins.unsafeGetAttrPos "hello" pkgs`. Counter-arguments:
+  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
+  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)
+
+# Alternatives
+[alternatives]: #alternatives
+
+## A different unit base directory
+
+Context: `pkgs/unit` is the unit base directory
+
+Alternatives:
+- Use `unit` in the root directory instead
+  - (+) This is future proof in case we want to make the directory structure more general purpose
+    - (-) We don't yet know if we want that, so this is out of scope for now
+- Use `pkgs` instead, so that the `${shard}`'s are siblings to the other current directories in `pkgs` such as `top-level`, with the intention that the other directories would be hopefully removed at some point, then only leaving the shards in `pkgs`
+  - (+) If we remove the other directories at some point, only the `${shard}`'s will be left in `pkgs`
+  - (-) This leads to ambiguities between the directories from the standard and the other directories, requiring special handling in the code and CI, leading to complexities.
+  - (-) This would be harder to document and explain to people, since one always has to disregard all non-unit directories, with no obvious justification
+  - (-) Currently we cannot apply this standard to all definitions in `pkgs`, in particular nested packages like `pythonPackages.*`, non-`callPackage`'d definitions like `copyDesktopItems` and non-derivations like `fetchFromGitHub`.
+    Depending on how we want to handle those, it might make more sense to keep `pkgs/unit` or to use `pkgs` directly once all legacy paths are migrated away to another top-level directory, we don't yet know. `pkgs/unit` will be easier to migrate to `pkgs` than the other way around though.
+- Another name than `unit`, proposed were `auto`, `pkg`, `mod`, `component`, `part`, `comp`, `app`, `simple`
+  - (-) `unit` is considered the most appropriate because a "unit" typically refers to a discrete, indivisible entity that is distict from other entities of the same type.
+  - (-) In addition we envision that in the future we would extend the unit directory standard to not just include a package definition for each unit, but also other parts such as NixOS modules, library components, tests, etc. In this case `unit` would fit even better and could be described as
+    > A collection of standardized files related to the same software component
```

Comment:
I don't think I follow the argument against `<nixpkgs>/pkgs/unit`. The way I read the RFC here, expanding the concept to modules in the future does not imply all of them should be put together: I could envisage a solution with `<nixpkgs>/modules/unit` for example.

In general, I quite like the top-level folder structure of nixpkgs and would prefer solutions that don't change it.

Reply:
> expanding the concept to modules in the future does not imply all of them should be put together

The term `unit` did originate from that idea, in the discussions we've had while drafting this RFC.

Putting related files together is a simplification that helps newcomers most of all. Preserving existing habits would not be a goal, but it would be a consideration.

The `<nixpkgs>/unit` idea is far more radical, so I think we were right to scope it out, but scoping it out does leave a loose end here. Not sure what to do, so I'm inclined to just deliver Simple Package Paths and save "Simple Paths" or whatever for a later RFC.
"Radical": I do think we want to get to the root at some point, figuratively and literally, but perhaps without the connotation.